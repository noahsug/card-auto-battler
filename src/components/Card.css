.Card {
  border: 1px solid #ccc;
  width: 24vh;
  height: 32vh;
  transition: transform .2s;
  margin: auto;
}

.Card-active {
  border: 5px solid #333;
  transform: scale(1.2);
}
