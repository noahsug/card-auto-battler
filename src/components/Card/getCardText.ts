import { assert, assertIsNonNullable, assertType } from '../../utils';
import {
  CardEffect,
  CardState,
  ValueDescriptor,
  MaybeValue,
  BasicValueDescriptor,
  PlayerValueDescriptor,
  If,
  Target,
} from '../../gameState/gameState';
import { KeysOfUnion } from '../../utils/types';

// Deal damage and gain strength equal to 2 times the enemy's bleed.

// const damageAndStrengthFromBleed = {
//   target: 'opponent',
//   effect: 'damage',
//   and: {
//     target: 'self',
//     effect: 'strength',
//   },
//   value: 2,
//   // times: 1, (1 by default)
//   multiplyBy: {
//     type: 'opponent',
//     name: 'bleed',
//   },
// };

// Deal 10 damage.
// Take 5 damage if this misses.

// const highJumpKick = {
//   effects: [{
//     target: 'opponent',
//     effect: 'damage',
//     value: 10,
//   }],
//   then: {
//     effects: [{
//       target: 'self',
//       effect: 'damage',
//       value: 5,
//       if: {
//         type: 'miss',
//       },
//     }],
//   },
// };

// Deal 10 damage.
// Misses if you have more HP than the enemy.

// const damageIfHigherHP = {
//   effects: [{
//     target: 'opponent',
//     effect: 'damage',
//     value: 10,
//   }],
//   modify: {
//     effect: 'miss',
//     if: {
//       type: 'self',
//       name: 'health',
//       comparison: '>',
//       compareTo: {
//         type: 'opponent',
//       },
//     },
//   },
// };

// Deal 5 damage.
// Deals double damage if you have less than half HP.

// const doubleDamageIfLowHp = {
//   effects: [{
//     target: 'opponent',
//     effect: 'damage',
//     value: 5,
//   }],
//   modify: {
//     effect: 'doubleDamage',
//     if: {
//       type: 'maxHealth',
//       target: 'self',
//       comparison: '<',
//       value: 'half',
//     },
//   },
// };

// Deal 5 damage.
// Deals 3 extra damage if you have more HP than bleed.

// Deal 3 damage.
// Deals extra damage for every 5 missing health.
// Apply bleed equal to damage dealt.

// const damageForMissingHealth = {
//   target: 'opponent',
//   effect: 'damage',
//   value: 3,
//   additionalEffect: {
//     target: 'opponent',
//     effect: 'damage',
//     multiplyBy: {
//       type: 'playerValue',
//       target: 'opponent',
//       name: 'missingHealth',
//       divisor: 5,
//     },
//   },
//   additionalEffect: {
//     target: 'opponent',
//     effect: 'bleed',
//     value: 1,
//   },
// };
// }

// Deal damage equal to the turn number.

// Deal damage and apply bleed equal to the number of cards you've played this turn.

// BAD: confusing, do not support
// Deal 3 damage and gain 2 strength if you have full HP.

type Translations = { [key: string]: () => string };

// note: We need getTranslationsList to be a function (rather that simply passing in the
// translations list) to avoid an infinite loop.
function getTranslationFn<T extends Translations>(getTranslationsList: () => T[]) {
  return (text: KeysOfUnion<T>): string => {
    const translations = getTranslationsList().find((t) => text in t);

    // return an empty string when an optional property is missing, e.g. when we see
    // t('if the enemy has more than 3 bleed') but `effect.if` is undefined
    if (!translations) return '';

    return translations[text]();
  };
}

function getNestedPlayerValueTranslations(playerValue: PlayerValueDescriptor) {
  return [getPlayerValueTranslations(playerValue), getTargetTranslations(playerValue.target)];
}

function getNestedValueTranslations(value: ValueDescriptor) {
  return value.type === 'basicValue'
    ? [getBasicValueTranslations(value)]
    : getNestedPlayerValueTranslations(value);
}

function getNestedIfTranslations(ifStatement?: If) {
  if (ifStatement == null) return [];
  return [
    getIfTranslations(ifStatement),
    ...getNestedPlayerValueTranslations(ifStatement.value),
    getBasicValueTranslations(ifStatement.value2),
  ];
}

function getNestedRepeatTranslations(repeat: MaybeValue) {
  return [
    getRepeatTranslations(repeat),
    ...getNestedValueTranslations(repeat.value),
    ...getNestedIfTranslations(repeat.if),
  ];
}

interface EffectOptions {
  addingToValue?: boolean;
}

function getNestedEffectTranslations(effect: CardEffect, options: EffectOptions = {}) {
  return [
    getEffectTranslations(effect, options),
    getTargetTranslations(effect.target),
    ...getNestedValueTranslations(effect.value),
    ...getNestedIfTranslations(effect.if),
  ];
}

function getTargetTranslations(target: Target) {
  return {
    [`you've`]: () => {
      return target === 'self' ? `you've` : `the enemy has`;
    },
    [`you have`]: () => {
      return target === 'self' ? `you have` : `the enemy has`;
    },
    [`your`]: () => {
      return target === 'self' ? `your` : `the enemy's`;
    },
    ['Enemy']: () => {
      return target === 'self' ? `You` : `Enemy`;
    },
  };
}

function getBasicValueTranslations({ value }: BasicValueDescriptor) {
  return {
    ['cards']: () => {
      return value === 1 ? 'card' : 'cards';
    },

    ['3']: () => {
      return String(value);
    },
  };
}

function getPlayerValueTranslations(playerValue: PlayerValueDescriptor) {
  const t = getTranslationFn(() => getNestedPlayerValueTranslations(playerValue));

  // not supported
  assert(playerValue.name != 'currentCardIndex');
  assert(playerValue.name != 'startingHealth');

  return {
    [`your bleed`]: (): string => {
      if (playerValue.name === 'trashedCards') {
        return `the number of cards ${t(`you've`)} trashed`;
      }
      if (playerValue.name === 'cardsPlayedThisTurn') {
        return `the number of cards ${t(`you've`)} played ${t('this turn')}`;
      }
      return `${t(`your`)} ${t('bleed')}`;
    },

    ['this turn']: (): string => {
      return playerValue.target === 'opponent' ? 'last turn' : 'this turn';
    },

    ['bleed']: () => {
      if (playerValue.name === 'health') return 'HP';
      return playerValue.name;
    },

    [`twice`]: () => {
      switch (playerValue.multiplier) {
        case undefined:
        case 1:
          return '';
        case 2:
          return 'twice';
        case 0.5:
          return 'half';
        default: {
          if (playerValue.multiplier > 1) {
            return `${playerValue.multiplier} times`;
          }
          // 1/4
          const denominator = (1 / playerValue.multiplier).toFixed(0);
          return `1/${denominator}`;
        }
      }
    },
  };
}

function getIfTranslations(ifStatement: If) {
  const t = getTranslationFn(() => getNestedIfTranslations(ifStatement));

  // not supported
  assert(ifStatement.value.multiplier == null);

  return {
    [`if the enemy has more than 3 bleed`]: (): string => {
      switch (ifStatement.value.name) {
        case 'trashedCards':
          return `if ${t(`you've`)} trashed ${t('more than 3')} cards`;

        case 'cardsPlayedThisTurn':
          return `if ${t(`you've`)} played ${t('more than 3')} cards this turn`;

        default:
          return `if ${t('you have')} ${t('more than 3')} ${t('bleed')}`;
      }
    },

    [`more than 3`]: (): string => {
      const { comparison, value2 } = ifStatement;
      const isCheckingExistence =
        (comparison === '>' && value2.value === 0) || (comparison === '>=' && value2.value === 1);
      if (isCheckingExistence) {
        // (if you have) "" (bleed)
        return '';
      }
      return `${t('more than')} ${t(`3`)}`;
    },

    [`more than`]: () => {
      switch (ifStatement.comparison) {
        case '>':
          return 'more than';
        case '<':
          return 'less than';
        case '=':
          return '';
        case '<=':
          return 'no more than';
        case '>=':
          return 'at least';
      }
      return ifStatement.comparison satisfies never;
    },
  };
}

function getEffectTranslations(effect: CardEffect, options: EffectOptions = {}) {
  const t = getTranslationFn(() => getNestedEffectTranslations(effect, options));

  return {
    [`Deal damage equal to your bleed`]: (): string => {
      switch (effect.value.type) {
        case 'playerValue':
          return `${t(`Deal damage`)} ${t(`equal to your bleed`)}`;
        case 'basicValue':
          return t(`Deal 3 damage`);
      }
      return effect.value satisfies never;
    },

    [`Deal damage`]: (): string => {
      if (effect.name === 'trash') {
        return `${t(`You trash`)} ${t('extra')}`;
      }
      if (effect.name === 'extraCardPlays') {
        return t('Enemy plays extra cards next turn');
      }
      return `${t(`Deal`)} ${t(`extra`)} ${t(`damage`)}`;
    },

    [`Deal 3 damage`]: (): string => {
      if (effect.name === 'trash') {
        return `${t(`You trash`)} ${t('3')} ${t('extra')}`;
      }
      if (effect.name === 'extraCardPlays') {
        return `${t('Enemy plays 3 extra cards next turn')}`;
      }
      return `${t(`Deal`)} ${t('3')} ${t(`extra`)} ${t(`damage`)}`;
    },

    ['Deal double damage']: (): string => {
      const { multiply } = effect;
      if (multiply == null) return '';

      // not supported - would need 'Apply twice as much poison if...'
      assert(effect.name === 'damage');

      return `${t('Deal')} ${t('double')} ${t(`damage`)}`;
    },

    ['Enemy plays 3 extra cards next turn']: (): string => {
      if (effect.target === 'self') {
        return `Play ${t('3')} ${t(`extra`)} ${t('cards')}`;
      }
      return `Enemy plays ${t('3')} extra ${t('cards')} next turn`;
    },

    ['Enemy plays extra cards next turn']: (): string => {
      if (effect.target === 'self') {
        return `Play ${t(`extra`)} cards`;
      }
      return `Enemy plays extra cards next turn`;
    },

    [`equal to your bleed`]: (): string => {
      return `equal to ${t('twice')} ${t('your bleed')}`;
    },

    [`You trash`]: () => {
      return effect.target === 'self' ? `You trash` : `Enemy trashes`;
    },

    [`Deal`]: () => {
      if (effect.name === 'damage') {
        return effect.target === 'self' ? 'Take' : 'Deal';
      }
      if (effect.target === 'self') {
        return 'Gain';
      }
      const isFriendlyEffect =
        effect.name === 'heal' || effect.name === 'strength' || effect.name === 'dodge';
      if (isFriendlyEffect) {
        return 'Enemy gains';
      }
      return 'Apply';
    },

    ['extra']: () => {
      return options.addingToValue ? 'extra' : '';
    },

    ['damage']: () => {
      if (effect.name === 'heal') return 'HP';
      return effect.name;
    },

    [`3 times`]: () => {
      if (effect.multiHit == null || effect.multiHit === 1) return '';
      return `${effect.multiHit} times`;
    },

    ['double']: () => {
      const basicValue = effect.multiply?.value;
      assertIsNonNullable(basicValue);

      // not supported - "Damage is reduced by 50%"
      assert(basicValue.value > 1);

      switch (basicValue.value) {
        case 2:
          return 'double';
        case 3:
          return 'triple';
        case 4:
          return 'quadruple';
        default: {
          const percent = (basicValue.value * 100).toFixed(0);
          return `${percent}%`;
        }
      }
    },
  };
}

function getRepeatTranslations(repeat: MaybeValue) {
  const t = getTranslationFn(() => getNestedRepeatTranslations(repeat));

  // not supported
  assertType(repeat.value, 'playerValue');
  assert(repeat.value.multiplier == null);
  assert(repeat.value.name != 'cardsPlayedThisTurn');
  assert(repeat.value.name != 'currentCardIndex');
  assert(repeat.value.name != 'extraCardPlays');
  assert(repeat.value.name != 'startingHealth');
  assert(repeat.value.name != 'trashedCards');

  return {
    [`Repeat for each bleed you have`]: (): string => {
      return `Repeat for each ${t('bleed')} ${t('you have')}`;
    },
  };
}

function getAddText(effect: CardEffect) {
  if (effect.add == null) return '';

  const addEffect: CardEffect = Object.assign({}, effect, {
    value: effect.add.value,
    if: effect.add.if,
  });
  const options = { addingToValue: true };

  const t = getTranslationFn(() => getNestedEffectTranslations(addEffect, options));
  return `${t('Deal damage equal to your bleed')} ${t('if the enemy has more than 3 bleed')}`;
}

function getCardEffectText(effect: CardEffect) {
  const t = getTranslationFn(() => getNestedEffectTranslations(effect));
  const effectText = [
    t(`Deal damage equal to your bleed`),
    t(`3 times`),
    t(`if the enemy has more than 3 bleed`),
  ].join(' ');

  const addText = getAddText(effect);

  const tm = getTranslationFn(() => getNestedIfTranslations(effect.multiply?.if));
  const multiplyText = `${t('Deal double damage')} ${tm('if the enemy has more than 3 bleed')}`;

  return [effectText, addText, multiplyText];
}

function getRepeatText(repeat?: MaybeValue) {
  if (!repeat) return [];

  const t = getTranslationFn(() => getNestedRepeatTranslations(repeat));
  const text = `${t(`Repeat for each bleed you have`)} ${t('if the enemy has more than 3 bleed')}`;

  return [text];
}

// Remove extra spaces
function fixSpacing(text: string) {
  return text.trim().replace(/\s+/g, ' ');
}

function isNotEmpty(text: string) {
  return !text.match(/^\s*$/);
}

export default function getCardText(card: CardState): string[] {
  return [...card.effects.flatMap(getCardEffectText), ...getRepeatText(card.repeat)]
    .filter(isNotEmpty)
    .map(fixSpacing);
}
