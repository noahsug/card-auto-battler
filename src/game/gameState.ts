import { EnemyName } from '../content/enemies';
import { heroesByName } from '../content/heroes';
import { getRandomState, Random } from '../utils/Random';
import { addCardsToPlayer } from './utils/cards';
import { getEnemyOrder } from './utils/enemies';
import { getNextEnemy } from './utils/selectors';

export type CardShopName = 'removeCards' | 'chainCards' | 'addPotions' | 'featherCards';
export type ShopName = CardShopName | 'addRelics';

export type Target = 'self' | 'opponent';

export const statusEffectNames = [
  'bleed',
  'extraCardPlays',
  'dodge',
  'temporaryDodge',
  'strength',
  'temporaryStrength',
  'regen',
  'burn',
  'lifesteal',
  'lifestealWhenBurning',
  'shock',
  'delayedShock',
  'stun',
  'crit',
  'temporaryFireCrit',
] as const;
export type StatusEffectName = (typeof statusEffectNames)[number];
export type StatusEffects = Record<StatusEffectName, number>;
export const EMPTY_STATUS_EFFECTS = Object.fromEntries(
  statusEffectNames.map((effectName) => [effectName, 0]),
) as StatusEffects;

type CalculatedPlayerValueName =
  | 'percentGreen'
  | 'percentRed'
  | 'percentPurple'
  | 'prevCardIsGreen'
  | 'prevCardIsRed'
  | 'prevCardIsPurple'
  | 'cardDamageDealtToTarget'
  | 'turn';

export type PlayerValueName =
  | keyof Omit<PlayerState, 'name' | 'image' | 'previousCard'>
  | CalculatedPlayerValueName;

// TODO: Add 'selfDamage' as a type, which isn't affected by things like strength
export type CardEffectName = StatusEffectName | 'damage' | 'heal' | 'trash';

export interface BasicValueDescriptor {
  type: 'basicValue';
  value: number;
}

export interface PlayerValueDescriptor {
  type: 'playerValue';
  target: Target;
  name: PlayerValueName;
  multiplier?: number;
}

export type ValueDescriptor = BasicValueDescriptor | PlayerValueDescriptor;

export interface If {
  value: PlayerValueDescriptor;
  comparison: '>' | '<' | '=' | '<=' | '>=';
  value2: BasicValueDescriptor;
}

export interface MaybeValue<T = ValueDescriptor> {
  value: T;
  if?: If;
}

export interface BasicCardEffect {
  target: Target;
  name: CardEffectName;
  value: ValueDescriptor;
  add?: MaybeValue;
  multiply?: MaybeValue<BasicValueDescriptor>;
  multiHit?: ValueDescriptor;
  if?: If;
}

export interface SetValueCardEffect extends Omit<BasicCardEffect, 'name'> {
  name: 'set';
  valueName: StatusEffectName | 'health';
}

export type CardEffect = BasicCardEffect | SetValueCardEffect;

export const tribes = ['basic', 'purple', 'red', 'green'] as const;
export type Tribe = (typeof tribes)[number];

export interface CardState {
  effects: CardEffect[];
  repeat?: MaybeValue;
  lifesteal?: MaybeValue;
  trash: boolean;
  uses?: { current: number; max: number };
  name: string;
  description: string;
  image: string;
  tribe: Tribe;
  // unique ID used to sort cards from first added to last added
  acquiredId: number;
  chain: { toId?: number; fromId?: number };
  charm?: 'feather';
}

export interface RelicState {
  name: string;
  displayName: string;
  description: string;
  value: number;
  value2: number;
  image: string;
  tribe: Tribe;
}

export interface PlayerState extends StatusEffects {
  health: number;
  startingHealth: number;
  cards: CardState[];
  trashedCards: CardState[];
  relics: RelicState[];
  currentCardIndex: number;
  cardsPlayedThisTurn: number;
  previousCard?: CardState;
  damageDealtThisTurn: number;
  damageDealtLastTurn: number;
  name: string;
  image: string;
  scale: number;
}

export interface GameState {
  user: PlayerState;
  enemy: PlayerState;
  enemyOrder: EnemyName[];
  turn: number;
  wins: number;
  losses: number;
  randomnessState: Uint32Array;
}

export interface PlayerInfo {
  name: string;
  image: string;
  cards: CardState[];
  health: number;
  scale: number;
  initialize?: (player: PlayerState) => void;
}

export function createPlayer({
  name,
  image,
  cards,
  health,
  scale,
  initialize,
}: PlayerInfo): PlayerState {
  const player = {
    health,
    startingHealth: health,
    ...EMPTY_STATUS_EFFECTS,
    cards: [],
    trashedCards: [],
    relics: [],
    currentCardIndex: 0,
    cardsPlayedThisTurn: 0,
    damageDealtThisTurn: 0,
    damageDealtLastTurn: 0,
    name,
    image,
    scale,
  };

  addCardsToPlayer(
    player,
    cards.map((c) => structuredClone(c)),
  );

  initialize?.(player);
  return player;
}

export function createGameState(seed?: number): GameState {
  const random = new Random(seed);
  const enemyOrder = getEnemyOrder(random);

  const game = {
    user: createPlayer(heroesByName.warrior),
    enemy: getNextEnemy({ enemyOrder, wins: 0 }),
    enemyOrder,
    turn: 0,
    wins: 0,
    losses: 0,
    randomnessState: getRandomState(seed),
  };

  // game.enemy.health = 5;
  // game.user.health = 1;

  return game;
}
