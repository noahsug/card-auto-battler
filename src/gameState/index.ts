export * from './gameState';
export * from './utils';
