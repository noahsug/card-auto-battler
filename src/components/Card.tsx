import styled, { css } from 'styled-components';

import { CardEffects, CardState, StatusEffects } from '../gameState';
import CardEffectText, { CardText } from './CardEffectText';

interface Props {
  card: CardState;
  isActive?: boolean;
  scale?: number;
  className?: string;
  onClick?: () => void;
}

function getCardTextItems(effects: CardEffects | undefined, targetSelf: boolean) {
  if (!effects) {
    return [];
  }

  const cardTextItems = [];

  if (effects.damage != null) {
    cardTextItems.push(<CardEffectText key="damage" effectName="damage" value={effects.damage} />);
  }

  const definedStatusEffects = effects.statusEffects
    ? Object.entries(effects.statusEffects).filter(([_, value]) => value != null)
    : [];

  definedStatusEffects.forEach(([statusEffectName, value]) => {
    cardTextItems.push(
      <CardEffectText
        key={`status-${statusEffectName}`}
        statusEffectName={statusEffectName as keyof StatusEffects}
        value={value!}
      />,
    );
  });

  if (effects.multihit != null) {
    cardTextItems.push(
      <CardEffectText key="multihit" effectName="multihit" value={effects.multihit} />,
    );
  }

  if (targetSelf) {
    cardTextItems.push(<CardText key="self">to self</CardText>);
  }

  return cardTextItems;
}

export default function Card({ card, isActive = false, scale = 1, className, onClick }: Props) {
  const targetTextItems = getCardTextItems(card.target, false);
  const selfTextItems = getCardTextItems(card.self, true);

  return (
    <Root $isActive={isActive} $scale={scale} className={className} onClick={onClick}>
      <CardTextSection>{targetTextItems}</CardTextSection>
      <CardTextSection>{selfTextItems}</CardTextSection>
    </Root>
  );
}

const Root = styled.div<{ $scale: number; $isActive: boolean }>`
  border: 1px solid #ccc;
  border-radius: 4px;
  padding: 10rem;
  transition:
    transform 0.2s,
    box-shadow 0.2s;

  width: ${({ $scale }) => $scale * 192}rem;
  height: ${({ $scale }) => $scale * 256}rem;

  ${({ $isActive }) =>
    $isActive &&
    css`
      transform: scale(1.2);
      box-shadow: 7rem 7rem 10rem 0 rgba(0, 0, 0, 0.5);
    `}
`;

const CardTextSection = styled.div`
  & + & {
    margin-top: 10rem;
  }
`;
