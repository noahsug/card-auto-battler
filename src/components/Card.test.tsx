import { render } from '@testing-library/react';

import Card from './Card';
import { CardState, createCard } from '../gameState';

describe('card text', () => {
  function getCardElement(cardState: CardState) {
    const { baseElement } = render(<Card card={cardState} />);
    return baseElement;
  }

  it('renders damage', () => {
    const card = getCardElement(createCard({ target: 'opponent', damage: 5 }));

    expect(card.textContent).toMatchInlineSnapshot(`"5‚öîÔ∏è"`);
  });

  it('renders heal', () => {
    const card = getCardElement(createCard({ target: 'self', heal: 5 }));

    expect(card.textContent).toMatchInlineSnapshot(`"5‚ù§Ô∏èto self"`);
  });

  it('renders self damage', () => {
    const card = getCardElement(createCard({ target: 'self', damage: 5 }));

    expect(card.textContent).toMatchInlineSnapshot(`"5‚öîÔ∏èto self"`);
  });

  describe('with activations', () => {
    it('renders positive activations', () => {
      const card = getCardElement(createCard({ target: 'opponent', damage: 5, activations: 3 }));

      expect(card.textContent).toMatchInlineSnapshot(`"5‚öîÔ∏è3x times"`);
    });
    it('renders 0 activations', () => {
      const card = getCardElement(createCard({ target: 'opponent', damage: 5, activations: 0 }));

      expect(card.textContent).toMatchInlineSnapshot(`"5‚öîÔ∏è0x times"`);
    });
  });

  it('renders multiple card effects', () => {
    const card = getCardElement(
      createCard(
        { target: 'opponent', damage: 5, bleed: 2 },
        { target: 'self', dodge: 1, extraCardPlays: 1, strength: 1, activations: 2 },
      ),
    );

    expect(card.textContent).toMatchInlineSnapshot(`"5‚öîÔ∏è2ü©∏1üÉè1üí®1üí™2x timesto self"`);
  });

  it('renders a status effect', () => {
    const card = getCardElement(createCard({ target: 'opponent', bleed: 2 }));

    expect(card.textContent).toMatchInlineSnapshot(`"2ü©∏"`);
  });

  it('renders trash self', () => {
    const card = getCardElement(createCard({ target: 'opponent', damage: 5, trashSelf: true }));

    expect(card.textContent).toMatchInlineSnapshot(`"5‚öîÔ∏ètrash this card"`);
  });

  it('renders trash', () => {
    const card = getCardElement(createCard({ target: 'opponent', damage: 3, trash: 2 }));

    expect(card.textContent).toMatchInlineSnapshot(`"3‚öîÔ∏è2trash"`);
  });

  it('renders conditionals', () => {
    const card = getCardElement(
      createCard({
        target: 'opponent',
        damage: 3,
        ifPlayerValue: {
          target: 'self',
          name: 'health',
          comparison: '<',
          compareToPercent: 50,
        },
      }),
    );

    expect(card.textContent).toMatchInlineSnapshot(`"3‚öîÔ∏èif self health < 50%"`);
  });

  describe('with effect gains', () => {
    it('renders +damage for every bleed', () => {
      const card = getCardElement(
        createCard({
          target: 'opponent',
          gainEffectsList: [
            {
              effects: { damage: 1 },
              forEveryPlayerValue: { target: 'opponent', name: 'bleed' },
            },
          ],
        }),
      );

      expect(card.textContent).toMatchInlineSnapshot(`"+1‚öîÔ∏è for every opponent ü©∏"`);
    });

    it('renders damage and +times for every strength', () => {
      const card = getCardElement(
        createCard({
          target: 'opponent',
          damage: 2,
          gainEffectsList: [
            {
              effects: { activations: 1 },
              forEveryPlayerValue: { target: 'self', name: 'strength' },
            },
          ],
        }),
      );

      expect(card.textContent).toMatchInlineSnapshot(`"2‚öîÔ∏è+1x times for every self üí™"`);
    });

    it('renders damage X times for every health', () => {
      const card = getCardElement(
        createCard({
          target: 'opponent',
          damage: 1,
          activations: 0,
          gainEffectsList: [
            {
              effects: { activations: 1 },
              forEveryPlayerValue: { target: 'self', name: 'health' },
            },
          ],
        }),
      );

      expect(card.textContent).toMatchInlineSnapshot(`"1‚öîÔ∏èfor every self hp"`);
    });

    it('renders +dodge for every trashed card', () => {
      const card = getCardElement(
        createCard({
          target: 'self',
          gainEffectsList: [
            {
              effects: { dodge: 1 },
              forEveryPlayerValue: { target: 'self', name: 'trashedCards' },
            },
          ],
        }),
      );

      expect(card.textContent).toMatchInlineSnapshot(`"+1üí® to self for every self trashed cards"`);
    });

    it('renders +heal for every damage dealt', () => {
      const card = getCardElement(
        createCard(
          { target: 'opponent', damage: 2 },
          {
            target: 'self',
            gainEffectsList: [
              {
                effects: { heal: 1 },
                forEveryBattleStat: { name: 'damageDealt' },
              },
            ],
          },
        ),
      );

      expect(card.textContent).toMatchInlineSnapshot(`"2‚öîÔ∏è+1‚ù§Ô∏è to self for every damage dealt"`);
    });
  });
});
