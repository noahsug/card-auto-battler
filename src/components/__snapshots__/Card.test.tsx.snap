// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`card text renders appliesStrengthTwiceCard 1`] = `"3⚔️+1⚔️ for every self 💪"`;

exports[`card text renders bleedCard 1`] = `"2🩸"`;

exports[`card text renders bleedStarterCard 1`] = `"1🩸"`;

exports[`card text renders bleedTrashCard 1`] = `"3🩸trash this card"`;

exports[`card text renders bothBleedCard 1`] = `"3🩸3🩸to self"`;

exports[`card text renders damageCard 1`] = `"5⚔️"`;

exports[`card text renders damageForEachBleedCard 1`] = `"1⚔️for every opponent 🩸"`;

exports[`card text renders damageForEachCard 1`] = `"+1⚔️ for every 2 self cards in decktrash this card"`;

exports[`card text renders damageForEachCardPlayedCard 1`] = `"3⚔️+2⚔️ for every self cards played this turn"`;

exports[`card text renders damageForEachMissingHealthCard 1`] = `"12⚔️+-1⚔️ for every 2 self health"`;

exports[`card text renders damageForEachTrashedCard 1`] = `"3⚔️+1⚔️ for every opponent trashed cards+1⚔️ for every self trashed cards"`;

exports[`card text renders damageSelfIfMissCard 1`] = `"7⚔️4⚔️to selfif hits = 0"`;

exports[`card text renders damageStarterCard 1`] = `"3⚔️"`;

exports[`card text renders damageTrashStarterCard 1`] = `"5⚔️trash this card"`;

exports[`card text renders dodgeAndTrashCard 1`] = `"2 trash2💨to selftrash this card2 trash2💨"`;

exports[`card text renders dodgeStarterCard 1`] = `"1💨to self"`;

exports[`card text renders doubleBleedCard 1`] = `"+1🩸, trash this card for every opponent 🩸"`;

exports[`card text renders doubleDodgeIfLowHealthCard 1`] = `"1💨+1💨 to self if self health <= 50% self max health"`;

exports[`card text renders doubleStrengthCard 1`] = `"+1💪, trash this card to self for every self 💪"`;

exports[`card text renders extraCardIfHighDamageCard 1`] = `"3⚔️+1🃏, +1💨 to self if damage dealt >= 7"`;

exports[`card text renders extraCardIfHighHealthCard 1`] = `"3❤️to self+2🃏 to self if self health >= self max health"`;

exports[`card text renders extraPlayCard 1`] = `"2⚔️1🃏to self"`;

exports[`card text renders extraPlayHealCard 1`] = `"3❤️1🃏to self"`;

exports[`card text renders extraPlayIfBleedCard 1`] = `"4⚔️+1🃏 to self if opponent 🩸 > 0"`;

exports[`card text renders extraPlayIfExtraPlayCard 1`] = `"1⚔️1🃏+1🃏 to self if self cards played this turn > 0"`;

exports[`card text renders extraPlayIfLowHealthCard 1`] = `"3⚔️+3🃏 to self if self health <= 50% self max health"`;

exports[`card text renders extraPlayStarterCard 1`] = `"1⚔️1🃏to self"`;

exports[`card text renders extraPlaysTrashCard 1`] = `"2🃏to selftrash this card"`;

exports[`card text renders gainStrengthForBleedCard 1`] = `"+1💪 to self for every opponent 🩸1🩸"`;

exports[`card text renders healCard 1`] = `"7❤️to self"`;

exports[`card text renders healForEachTrashedCard 1`] = `"4❤️+1❤️ to self for every opponent trashed cards+1❤️ to self for every self trashed cards"`;

exports[`card text renders healStarterCard 1`] = `"4❤️to self"`;

exports[`card text renders healTrashStarterCard 1`] = `"7❤️to selftrash this card"`;

exports[`card text renders lifestealCard 1`] = `"3⚔️+1❤️ to self for every damage dealt"`;

exports[`card text renders multihitCard 1`] = `"1⚔️3x times"`;

exports[`card text renders multihitStarterCard 1`] = `"1⚔️2x times"`;

exports[`card text renders plusHealForEachTrashedCard 1`] = `"4❤️+1❤️ to self for every opponent trashed cards"`;

exports[`card text renders selfDamageCard 1`] = `"7⚔️4⚔️to selfif self health > 50% self max health"`;

exports[`card text renders setHealthToHalfCard 1`] = `"1🃏+1⚔️ to self for every self health+1❤️ to self for every 2 self max health"`;

exports[`card text renders strengthCard 1`] = `"2💪to self"`;

exports[`card text renders strengthStarterCard 1`] = `"1💪to self"`;

exports[`card text renders strengthTrashCard 1`] = `"3💪to selftrash this card"`;

exports[`card text renders trashAndExtraPlayCard 1`] = `"2 trash2🃏to self"`;

exports[`card text renders trashAndTrashSelfCard 1`] = `"3 trashtrash this card"`;

exports[`card text renders trashCard 1`] = `"2 trash"`;

exports[`card text renders trashForOpponentHealthCard 1`] = `"+1 trash for every 5 opponent health"`;
